# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character：角色姓名。
- role：演职员类型，演员或导演。

# Read data.

In [1]:
import pandas as pd

In [2]:
ori_credit = pd.read_csv("E:\\01_Python\\ex_organize_data\\credits.csv")
ori_title = pd.read_csv("E:\\01_Python\\ex_organize_data\\titles.csv")

In [3]:
ori_credit.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

In [4]:
ori_title.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

# Evaluate and clean data.

In [5]:
clean_credit = ori_credit.copy()
clean_title = ori_title.copy()

# Evaluate data tidiness.

In [6]:
ori_title.head(5)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [7]:
pd.set_option('display.max_colwidth', 1000)
ori_title.head(4)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                                                                                                                                                                                                                                                                                                                                                                                                       description  \
0                                                                                                                                                                                                                                                                          This collection includes 12 World War II-era propaganda films — many of which are graphic and offensive — discussed in the docuseries "Five Came Back."   
1                                                                                                                                                                                                                                                            A mentally unstable Vietnam War veteran works as a night-time taxi driver in New York City where the perceived decadence and sleaze feed his urge for violent action.   
2                                                                                                                                                                                                            Intent on seeing the Cahulawassee River before it's turned into one huge lake, outdoor fanatic Lewis Medlock takes his friends on a river-rafting trip they'll never forget into the dangerous American back-country.   
3  King Arthur, accompanied by his squire, recruits his Knights of the Round Table, including Sir Bedevere the Wise, Sir Lancelot the Brave, Sir Robin the Not-Quite-So-Brave-As-Sir-Lancelot and Sir Galahad the Pure. On the way, Arthur battles the Black Knight who, despite having had all his limbs chopped off, insists he can still fight. They reach Camelot, but Arthur decides not  to enter, as "it is a silly place".   

   release_year age_certification  runtime  \
0          1945             TV-MA       51   
1          1976                 R      114   
2          1972                 R      109   
3          1975                PG       91   

                                        genres production_countries  seasons  \
0                            ['documentation']               ['US']      1.0   
1                           ['drama', 'crime']               ['US']      NaN   
2  ['drama', 'action', 'thriller', 'european']               ['US']      NaN   
3              ['fantasy', 'action', 'comedy']               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811

## clean_title

在title里面，Inconsistent data might be in the `genres` and `production_countries` columns.

对genres这一列进行拆分成多行的操作

In [8]:
clean_title['genres'] = clean_title['genres'].apply(eval)


In [9]:
clean_title = clean_title.explode('genres')

In [10]:
clean_title['genres'].head(10)

0    documentation
1            drama
1            crime
2            drama
2           action
2         thriller
2         european
3          fantasy
3           action
3           comedy
Name: genres, dtype: object

`production_countries` is the exact same worlflow.

In [11]:
clean_title['production_countries'] = clean_title['production_countries'].apply(eval)

In [12]:
clean_title = clean_title.explode('production_countries')

In [13]:
clean_title['production_countries'].head(8)

0    US
1    US
1    US
2    US
2    US
2    US
2    US
3    GB
Name: production_countries, dtype: object

## clean_credit

In [14]:
clean_credit.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

There is no structural issues.

# Data cleanliness.

## clean_title

In [15]:
clean_title.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64

From the output, there are a total of 17,818 observations.  
`title`、` description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、 `imdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score` variable all have missing values.  
We need to evaluate and clean later.

`release_year`：发布年份，should be 'datetime'.

In [16]:
clean_title['release_year'] = pd.to_datetime(clean_title['release_year'], format='%Y')
clean_title['release_year']

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

## clean_credit

In [17]:
clean_credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


From the output, there are a total of 77,801 observations.  
`character` variable have missing values  
We need to evaluate and clean later.  

In [18]:
clean_credit['person_id'].head(8)

0      3748
1     14658
2      7064
3      3739
4     48933
5     32267
6    519612
7     29068
Name: person_id, dtype: int64

### Assess missing data.

在`clean_title`里面     
`title`、影视作品标题    
` description`、简短描述    
`age_certification`、适龄认证    
`genres`、流派类型列表    
`production_countries`、出品国家列表   
`seasons`、季数    
`imdb_id`、IMDB的ID    
`imdb_score`、IMDB的评分    
`imdb_votes`、IMDB的投票数    
`tmdb_popularity`、TMDB的流行度    
`tmdb_score`TMDB的评分   
variable all have missing values.  

我们的分析目的是挖掘各个流派中的高TMDB评分作品的演员，so`title`，` description`，`age_certification`，`production_countries`，`seasons`可以保留空缺值，  
Besides, Lily said  `imdb_id`、`imdb_votes`、`tmdb_popularity`、`tmdb_score` these variables don't matter, so I follow my teacher Lily's word.

but，`imdb_score`IMDB评分    
`genres`流派    
is closely related to the analyze later.

先提取出`imdb_score`缺失观察值进行查看

In [19]:
clean_title.query('imdb_score.isnull()')

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                                                                                                                                                                                                                                                                                                    description  \
0                                                                                                                                                                       This collection includes 12 World War II-era propaganda films — many of which are graphic and offensive — discussed in the docuseries "Five Came Back."   
75    Sane Man was filmed before Bill recorded ‘Dangerous’, his first comedy album, and is a turning point in Hicks’ career. It was the first complete Hicks show ever filmed and Bill pulled out all the stops for the cameras. Completely focused, a newly-sober Hicks paces the stage like a wild animal riffing effortless.   
145                                                                                                                                                                                         “Hajimete no Otsukai” (First Errand) is a Japanese TV show that follows small children as they carry out errands for their parents.   
145                                                                                                                                                                                         “Hajimete no Otsukai” (First Errand) is a Japanese TV show that follows small children as they carry out errands for their parents.   
145                                                                                                                                                                                         “Hajimete no Otsukai” (First Errand) is a Japanese TV show that follows small children as they carry out errands for their parents.   
...                                                                                                                                                                                                                                                                                                                         ...   
5810                                                                                                                                                                       Two competitive social media Influencers go head-to-head to snatch the "Social Man" of the year title and are willing to break all the rules to win.   
5833                                                                                                                                                                             When a family run barber shop in the heart of Lagos is threatened by real estate developers, they'll do whatever it takes to stay in business.   
5840                                                                              In 14th-century Mali, an ambitious young royal named Mansa Musa ascended the throne of the richest kingdom in human history. This follows Malian artist Abdou Ouologuem on a journey to discover th

把这些观察值delete，之后查看delete之后该列空缺值个数的和进行验证

In [20]:
clean_title = clean_title.dropna(subset=['imdb_score'])
clean_title['imdb_score'].isnull().sum()

0

再提取出`genres`缺失观察值进行查看

In [21]:
clean_title.query('genres.isnull()')

id                                                     title  \
1813    ts77824  My Next Guest Needs No Introduction With David Letterman   
1939   ts215037                                     Minecraft: Story Mode   
2386    ts74805                          A Little Help with Carol Burnett   
2658   ts265844                                                  #ABtalks   
4274  tm1172010                                         The Lockdown Plan   
4648  tm1113921                                                  In Vitro   

       type  \
1813   SHOW   
1939   SHOW   
2386   SHOW   
2658   SHOW   
4274  MOVIE   
4648  MOVIE   

                                                                                                                                                                                                 description  \
1813                                                   TV legend David Letterman teams up with fascinating global figures for in-depth interviews and curiosity-fueled excursions in this monthly talk show.   
1939  MInecraft: Story Mode is an interactive, animated TV series from Netflix and Telltale Games. Viewers input decisions that affect the course of the story, using a remote control or computer keyboard.   
2386                                                    In this unscripted series starring comedy legend Carol Burnett, kids dish out advice to celebrities and everyday people in front of a live audience.   
2658                                        #ABtalks is a YouTube interview show hosted by Anas Bukhash, showcasing the 'raw' side of celebrities, athletes, entrepreneurs, and influencers as human beings.   
4274                                                                                                                                                                                                     NaN   
4648                                                                         'In Vitro' is an otherworldly rumination on memory, history, place and identity set in Bethlehem decades after an eco-disaster.   

     release_year age_certification  runtime genres production_countries  \
1813   2018-01-01             TV-MA       50    NaN                   US   
1939   2018-01-01             TV-PG       52    NaN                   US   
2386   2018-01-01              TV-G       24    NaN                   US   
2658   2018-01-01             TV-PG       68    NaN                  NaN   
4274   2020-01-01               NaN       49    NaN                  NaN   
4648   2019-01-01               NaN       27    NaN                  NaN   

      seasons     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813      4.0   tt7829834         7.8      5581.0            8.217         7.6  
1939      1.0  tt10498322         5.6       347.0              NaN         NaN  
2386      1.0   tt7204366         6.3       237.0            1.621         6.2  
2658      1.0  tt12635254         9.6         7.0              NaN         NaN  
4274      NaN  tt13079112         6.5         NaN              NaN         NaN  
4648      NaN  tt10545994         7.7         NaN              NaN         NaN

In [22]:
clean_title = clean_title.dropna(subset=['genres'])
clean_title['genres'].isnull().sum()

0

`clean_credit`中，    
`character`variable：角色姓名     
have missing values.

`character` has nothing to do with our analyze target, so we can keep this missing value.

### Handle duplicate data.

In [23]:
clean_title.duplicated().sum()

0

In [24]:
clean_credit.duplicated().sum()

0

There is no duplicate data in both them.

### Handle inconsistent data.

### clean_title

Inconsistent data might be in the `genres`和`production_countries` variables.Next, verify.

In [25]:
clean_title['genres'].value_counts()

drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: genres, dtype: int64

From the output, there is no inconsisstent data in the `genres`.

### clean_title

In [26]:
clean_title['production_countries'].value_counts()

US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: production_countries, Length: 108, dtype: int64

In [27]:
with pd.option_context('display.max_rows', None):
    print(clean_title['production_countries'].value_counts())

US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK            9
MT      

From the output，`LB` and `Lebanon`表示同一国家，need to standardize them.

So, `clean_title`里，make sure `"LB"` and `"Lebanon"` are both recorded as `LB` in `production_countries`, then check the results.

In [28]:
clean_title['production_countries'] = clean_title['production_countries'].replace( 'Lebanon', 'LB')
with pd.option_context('display.max_rows', None):
    print(clean_title['production_countries'].value_counts())

US    5648
IN    1610
GB    1068
JP    1046
FR     720
KR     637
ES     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
AT      28
SA      28
IS      28
LU      27
NZ      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
UY      15
MA      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
MT       8
VN       8
LT       7
IR       7
CD       7
SU       7
TN       7
SN       6
AL       6
KE       6
GH       6
IQ       5
MU       5
CY       5
KN       4
GR       4
IO       4
SY       4
TZ       4
MC       4
GL       3
CM       3
AO       3
BS       3
HR       3
BD       3
PY       3
DZ       3

### clean_credit

Inconsistent data might be in the `role`, we need to assess.

In [29]:
clean_credit['role'].value_counts()

ACTOR       73251
DIRECTOR     4550
Name: role, dtype: int64

From the output, there is no inconsistent data in the `role`.

我们可以把这列的类型转换为`Category`，好处是比字符串类型更节约内存空间，也能表明说值的类型有限。

In [30]:
clean_credit['role'] = clean_credit['role'].astype('category')
clean_credit['role']

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

## Handle invalid or incorrect data.

### clean_title

In [31]:
clean_title.describe()

runtime      seasons    imdb_score    imdb_votes  tmdb_popularity  \
count  16970.000000  5954.000000  16970.000000  1.694100e+04     16842.000000   
mean      80.912552     2.455492      6.514207  3.281655e+04        29.396307   
std       39.596172     2.869428      1.131095  1.141492e+05        93.178235   
min        0.000000     1.000000      1.500000  5.000000e+00         0.600000   
25%       45.000000     1.000000      5.800000  7.800000e+02         4.070000   
50%       90.000000     2.000000      6.600000  3.508000e+03        10.195000   
75%      107.000000     3.000000      7.300000  1.697800e+04        23.639000   
max      225.000000    42.000000      9.500000  2.294231e+06      2274.044000   

         tmdb_score  
count  16515.000000  
mean       6.846933  
std        1.078831  
min        1.000000  
25%        6.200000  
50%        6.900000  
75%        7.500000  
max       10.000000

Based on the states above, there are no unrealistic values in`original_titles`.

### clean_credit

不包含表示数值的变量，无需检查

# Organize data.

In [32]:
clean_title

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                                                                                                                                                                                                description  \
1                                                     A mentally unstable Vietnam War veteran works as a night-time taxi driver in New York City where the perceived decadence and sleaze feed his urge for violent action.   
1                                                     A mentally unstable Vietnam War veteran works as a night-time taxi driver in New York City where the perceived decadence and sleaze feed his urge for violent action.   
2     Intent on seeing the Cahulawassee River before it's turned into one huge lake, outdoor fanatic Lewis Medlock takes his friends on a river-rafting trip they'll never forget into the dangerous American back-country.   
2     Intent on seeing the Cahulawassee River before it's turned into one huge lake, outdoor fanatic Lewis Medlock takes his friends on a river-rafting trip they'll never forget into the dangerous American back-country.   
2     Intent on seeing the Cahulawassee River before it's turned into one huge lake, outdoor fanatic Lewis Medlock takes his friends on a river-rafting trip they'll never forget into the dangerous American back-country.   
...                                                                                                                                                                                                                     ...   
5846                                                                                          A heart warming film that explores the concept of romance with episodes involving four sets of people of different age group.   
5847                                                                A controversial TV host and comedian who has built his career on sexist humor is forced to assume a woman's identity to elude a relentless drug dealer.   
5849                                                                         With winter behind them, Bheem and his townspeople usher in a sunny new season in all their favorite ways during the Makar Sankranti festival.   
5849                                                                         With winter behind them, Bheem and his townspeople usher in a sunny new season in all their favorite ways during the Makar Sankranti festival.   
5849                                                                         With winter behind them, Bheem and his townspeople usher in a sunny new season in all their favorite ways during the Makar Sankranti festival.   

     release_year age_certification  runtime     genres production_countries  \
1      1976-01-01                 R      114      drama                   US   
1      1976-01-01                 R      114      crime                   US   
2      1972-01-01                 R      109      drama                   US   
2      1972-01-01                 R      109     action                   US   
2      1972-01-01                 R      109   thriller                   US   
...           ...               ...      ...        ...                  ...   
5846   2021-01-01              

In [33]:
clean_credit

person_id         id                name                character  \
0           3748    tm84618      Robert De Niro            Travis Bickle   
1          14658    tm84618        Jodie Foster            Iris Steensma   
2           7064    tm84618       Albert Brooks                      Tom   
3           3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4          48933    tm84618     Cybill Shepherd                    Betsy   
...          ...        ...                 ...                      ...   
77796     736339  tm1059008    Adelaida Buscato                María Paz   
77797     399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798     373198  tm1059008         Inés Prieto                    Fanny   
77799     378132  tm1059008        Isabel Gaona                   Cacica   
77800    1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

此数据分析的目的是，整理不同流派影视作品中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。


In [34]:
credit_with_title = pd.merge(clean_credit, clean_title, on='id', how='inner')

In [35]:
pd.reset_option('display.max_colwidth')
credit_with_title.head()

person_id       id            name      character   role        title  \
0       3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
1       3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
2      14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
3      14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
4       7064  tm84618   Albert Brooks            Tom  ACTOR  Taxi Driver   

    type                                        description release_year  \
0  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  crime                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  crime                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.2    808582.0           40.965       8.179  
1         8.2    808582.0           40.965       8.179  
2         8.2    808582.0           40.965       8.179  
3         8.2    808582.0           40.965       8.179  
4         8.2    808582.0           40.965       8.179

由于我们挖掘演员的参演作品口碑，导演不在我们的分析范围内

In [36]:
actor_with_title = credit_with_title.query('role == "ACTOR"')

先根据演员和流派进行分组，actor和genres

In [37]:
groupby_genres_and_person_id = actor_with_title.groupby(['genres', 'person_id'])

提取出`imdb_score`进行求平均值来分析各个流派的影视作品中，每位actor的average IMDB score

In [38]:
imdb_score_groupby_genres_and_person_id = groupby_genres_and_person_id['imdb_score'].mean()

In [39]:
imdb_score_groupby_genres_and_person_id

genres   person_id
action   45           5.0
         48           5.4
         51           6.4
         53           6.8
         54           5.3
                     ... 
western  2353339      6.9
         2370848      6.1
         2398539      3.8
         2406218      6.0
         2408082      7.3
Name: imdb_score, Length: 168881, dtype: float64

把他变为更加规整的DataFrame

In [40]:
imdb_score_groupby_genres_and_person_id_df = imdb_score_groupby_genres_and_person_id.reset_index()

In [41]:
imdb_score_groupby_genres_and_person_id_df

genres  person_id  imdb_score
0        action         45         5.0
1        action         48         5.4
2        action         51         6.4
3        action         53         6.8
4        action         54         5.3
...         ...        ...         ...
168876  western    2353339         6.9
168877  western    2370848         6.1
168878  western    2398539         3.8
168879  western    2406218         6.0
168880  western    2408082         7.3

[168881 rows x 3 columns]

# -----------------------------------------------------------------------------------------

# 现在针对流派和演员分组的IMDB评分数据已经整理好，可以进入后续的分析步骤了,但是为了更进一步熟练数据整理，将要进行更加深入的数据整理环节

对上面的结果再次进行分组，找出各个流派里演员作品最高的平均评分是多少、最高评分对应的演员名字是什么

In [42]:
genres_max_score = imdb_score_groupby_genres_and_person_id_df.groupby('genres')['imdb_score'].max()
genres_max_score

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

In [56]:
genres_max_score_with_person_id = pd.merge(genres_max_score, imdb_score_groupby_genres_and_person_id_df, on=['genres', 'imdb_score'], how='inner')
genres_max_score_with_person_id

genres  imdb_score  person_id
0     action         9.3       1303
1     action         9.3      12790
2     action         9.3      21033
3     action         9.3      86591
4     action         9.3     336830
..       ...         ...        ...
131      war         8.8     826547
132  western         8.9      22311
133  western         8.9      28166
134  western         8.9      28180
135  western         8.9      93017

[136 rows x 3 columns]

增加`name`变量，从而展示平均评分最高的演员名字

In [55]:
actor_id_with_names = clean_credit[['person_id', 'name']].drop_duplicates()
actor_id_with_names.head(10)

person_id             name
0       3748   Robert De Niro
1      14658     Jodie Foster
2       7064    Albert Brooks
3       3739    Harvey Keitel
4      48933  Cybill Shepherd
5      32267      Peter Boyle
6     519612   Leonard Harris
7      29068   Diahnne Abbott
8     519613      Gino Ardito
9       3308  Martin Scorsese

In [59]:
genres_max_score_with_actor_name = pd.merge(
    genres_max_score_with_person_id,
    actor_id_with_names,
    on='person_id',
    how='inner'
)
genres_max_score_with_actor_name

genres  imdb_score  person_id                name
0       action         9.3       1303       Jessie Flower
1    animation         9.3       1303       Jessie Flower
2       family         9.3       1303       Jessie Flower
3      fantasy         9.3       1303       Jessie Flower
4        scifi         9.3       1303       Jessie Flower
..         ...         ...        ...                 ...
131        war         8.8     826547         Yuto Uemura
132    western         8.9      22311     Koichi Yamadera
133    western         8.9      28166  Megumi Hayashibara
134    western         8.9      28180      Unsho Ishizuka
135    western         8.9      93017            Aoi Tada

[136 rows x 4 columns]

把相同的genres都排列在一起

In [60]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values('genres').reset_index().drop('index', axis=1)
genres_max_score_with_actor_name

genres  imdb_score  person_id                name
0     action         9.3       1303       Jessie Flower
1     action         9.3      86591       Cricket Leigh
2     action         9.3      21033          Zach Tyler
3     action         9.3      12790         Olivia Hack
4     action         9.3     336830     André Sogliuzzo
..       ...         ...        ...                 ...
131      war         8.8     826547         Yuto Uemura
132  western         8.9      28166  Megumi Hayashibara
133  western         8.9      28180      Unsho Ishizuka
134  western         8.9      22311     Koichi Yamadera
135  western         8.9      93017            Aoi Tada

[136 rows x 4 columns]

# 数据整理完成啦，接下来就可以进行数据分析啦！    
# 撒花！！！！！！